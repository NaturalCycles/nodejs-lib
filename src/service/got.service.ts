import { GotJSONOptions } from 'got'
import * as got from 'got'
import * as http from 'http'
import { getDebug } from '../util/debug'
const debug = getDebug(__filename)

export interface GotOptions extends GotJSONOptions {}

export interface GotResponse<T = string> extends http.IncomingMessage {
  body: T
  url: string
  requestUrl: string
  fromCache: boolean
  redirectUrls?: string[]
}

class GotService {
  /**
   * Just a proxy method, to simplify importing `* as got`, etc.
   */
  async got<T = string> (url: string, opt: got.GotOptions<any> = {}): Promise<GotResponse<T>> {
    return got(url, opt) as any
  }

  async request<T = any> (method: string, url: string, _opt: Partial<GotOptions> = {}): Promise<T> {
    debug(`>> ${method} ${url}`)
    const opt: GotOptions = {
      method,
      json: true,
      ..._opt,
    }

    try {
      const r = await got(url, opt)
      debug(`<< ${method} ${url}`)
      return r.body
    } catch (err) {
      return this.handleError(err, method, url)
    }
  }

  private handleError (err: any, method: string, url: string): never {
    if (!err || !err.response) throw err

    const r = err.response
    const errBody = JSON.stringify(r.body || '<empty_body>', null, 2)
    const msg = `<< ERROR ${r.statusCode} ${method} ${url}\n${errBody}`
    const e: any = new Error(msg)
    e.response = r
    throw e
  }

  // convenience methods
  async get<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('GET', url, opt)
  }
  async post<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('POST', url, opt)
  }
  async put<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('PUT', url, opt)
  }
  async delete<T = any> (url: string, opt?: Partial<GotOptions>): Promise<T> {
    return this.request<T>('DELETE', url, opt)
  }
}

export const gotService = new GotService()
