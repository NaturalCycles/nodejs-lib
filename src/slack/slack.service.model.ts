import { StringMap } from '@naturalcycles/js-lib'
import { DebugLogLevel, InspectAnyOptions } from '..'

/**
 * Properties that exists both in SlackApiBody (as per Slack API) and SlackMessage (our abstraction).
 */
export interface SlackMessageProps {
  /**
   * @default bot
   */
  username?: string

  channel?: string
  icon_url?: string

  /**
   * @default :spider_web:
   */
  icon_emoji?: string

  attachments?: SlackMessageAttachment[]
}

export interface SlackApiBody extends SlackMessageProps {
  text: string
}

export interface SlackMessage<CTX = any> extends SlackMessageProps {
  /**
   * The only *required* field.
   *
   * You can throw anything at it, it'll handle it appropriately:
   * String - as is
   * Object - pass via util.inspect()
   * Array - will pass each item via util.inspect() and join with \n
   * Error - print the stack nicely
   *
   * If you don't want the default Array behavior - you can pre-util.inspect() it yourself to your liking.
   */
  items: any

  /**
   * Optional "context object", to be used by `messagePrefixHook`.
   */
  ctx?: CTX

  level?: DebugLogLevel

  /**
   * Keys-values will be rendered as MessageAttachment with Fields
   */
  kv?: StringMap<any>

  /**
   * If specified - adds @name1, @name2 in the end of the message
   */
  mentions?: string[]

  /**
   * @default false
   * By default it ignores possible errors from slack
   */
  throwOnError?: boolean

  /**
   * @default false
   * Skips logging message
   */
  noLog?: boolean

  /**
   * Defaults to:
   * includeErrorData: true
   * includeErrorStack: true
   */
  inspectOptions?: InspectAnyOptions
}

export interface SlackAttachmentField {
  title: string
  value: string
  short?: boolean
}

// Taken from here: https://github.com/slackapi/node-slack-sdk/blob/master/packages/types/src/index.ts
export interface SlackMessageAttachment {
  // blocks?: (KnownBlock | Block)[];
  fallback?: string // either this or text must be defined
  color?: 'good' | 'warning' | 'danger' | string
  pretext?: string
  author_name?: string
  author_link?: string // author_name must be present
  author_icon?: string // author_name must be present
  title?: string
  title_link?: string // title must be present
  text?: string // either this or fallback must be defined
  fields?: SlackAttachmentField[]
  image_url?: string
  thumb_url?: string
  footer?: string
  footer_icon?: string // footer must be present
  ts?: number
  // actions?: AttachmentAction[];
  callback_id?: string
  mrkdwn_in?: ('pretext' | 'text' | 'fields')[]
}

/**
 * Return `null` to skip (filter out) the message completely.
 */
export type SlackMessagePrefixHook<CTX = any> = (
  msg: SlackMessage<CTX>,
) => string[] | null | Promise<string[] | null>

export interface SlackServiceCfg<CTX = any> {
  /**
   * Undefined means slack is disabled.
   */
  webhookUrl?: string

  defaults?: Partial<SlackMessage>

  /**
   * Override channel when msg.level is set.
   * key: DebugLogLevel
   * value: channel name to send message to
   */
  channelByLevel?: StringMap

  /**
   * Function to return an array of "prefix tokens" (will be joined by ': ').
   * Allows to skip (filter out) the message by returning `null`.
   */
  messagePrefixHook: SlackMessagePrefixHook<CTX>
}
