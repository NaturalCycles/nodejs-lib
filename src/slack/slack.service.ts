import {
  _omit,
  AnyObject,
  CommonLogger,
  commonLoggerMinLevel,
  CommonLogLevel,
  PQueue,
} from '@naturalcycles/js-lib'
import { dayjs } from '@naturalcycles/time-lib'
import got from 'got'
import { inspectAny, InspectAnyOptions } from '..'
import {
  SlackApiBody,
  SlackAttachmentField,
  SlackMessage,
  SlackServiceCfg,
} from './slack.service.model'

const GAE = !!process.env['GAE_INSTANCE']

const DEFAULTS: SlackMessage = {
  username: 'bot',
  channel: '#log',
  icon_emoji: ':spider_web:',
  items: 'no text',
}

const INSPECT_OPT: InspectAnyOptions = {
  colors: false,
  includeErrorData: true,
  includeErrorStack: true,
}

/**
 * Has 2 main methods:
 *
 * 1. .send({ items: any, channel: ..., ... })
 * Low-level method with all possible options available.
 *
 * 2. .log(...items: any[])
 * Shortcut method to "just log a bunch of things", everything is "by default" there.
 *
 * .send method has a shortcut:
 * .send(string, ctx?: CTX)
 */
export class SlackService<CTX = any> {
  constructor(cfg: Partial<SlackServiceCfg<CTX>>) {
    this.cfg = {
      messagePrefixHook: slackDefaultMessagePrefixHook,
      logger: console,
      ...cfg,
      inspectOptions: {
        ...INSPECT_OPT,
        ...cfg.inspectOptions,
      },
    }
  }

  public cfg!: SlackServiceCfg<CTX>

  /**
   * Allows to "log" many things at once, similar to `console.log(one, two, three).
   */
  async log(...items: any[]): Promise<void> {
    await this.send({
      // If it's an Array of just 1 item - treat it as non-array
      items: items.length === 1 ? items[0] : items,
    })
  }

  async send(input: SlackMessage<CTX> | string, ctx?: CTX): Promise<void> {
    const { webhookUrl, messagePrefixHook, inspectOptions } = this.cfg

    // If String is passed as first argument - just transform it to a full SlackMessage
    const msg = typeof input === 'string' ? { items: input } : input

    if (ctx !== undefined) {
      Object.assign(msg, { ctx })
    }

    this.cfg.logger.log(...[msg.items, msg.kv, msg.attachments, msg.mentions].filter(Boolean))

    if (!webhookUrl) return

    // Transform msg.kv into msg.attachments
    if (msg.kv) {
      ;(msg.attachments ||= []).push({ fields: this.kvToFields(msg.kv) })

      delete msg.kv // to not pass it all the way to Slack Api
    }

    let text: string

    // Array has a special treatment here
    if (Array.isArray(msg.items)) {
      text = msg.items.map(t => inspectAny(t, inspectOptions)).join('\n')
    } else {
      text = inspectAny(msg.items, inspectOptions)
    }

    // Wrap in markdown-text-block if it's anything but plain String
    if (typeof msg.items !== 'string') {
      text = '```' + text + '```'
    }

    if (msg.mentions?.length) {
      text += '\n' + msg.mentions.map(s => `<@${s}>`).join(' ')
    }

    const prefix = await messagePrefixHook(msg)
    if (prefix === null) return // filtered out!

    const json: SlackApiBody = _omit(
      {
        ...DEFAULTS,
        ...this.cfg.defaults,
        ...msg,
        // Text with Prefix
        text: [prefix.join(': '), text].filter(Boolean).join('\n'),
      },
      ['items', 'ctx'],
    )

    await got
      .post(webhookUrl, {
        json,
        responseType: 'text',
        timeout: 90_000,
      })
      .catch(err => {
        // ignore (unless throwOnError is set)
        if (msg.throwOnError) throw err
      })
  }

  kvToFields(kv: AnyObject): SlackAttachmentField[] {
    return Object.entries(kv).map(([k, v]) => ({
      title: k,
      value: String(v),
      short: String(v).length < 80,
    }))
  }

  /**
   * Returns a CommonLogger implementation based on this SlackService instance.
   */
  getCommonLogger(opt: {
    minLogLevel: CommonLogLevel
    logChannel?: string
    warnChannel?: string
    errorChannel?: string
  }): CommonLogger {
    const { minLogLevel = 'log', logChannel, warnChannel, errorChannel } = opt
    const defaultChannel = this.cfg.defaults?.channel || DEFAULTS.channel!

    const q = new PQueue({
      concurrency: 1,
    })

    return commonLoggerMinLevel(
      {
        log: (...args) =>
          q.push(() => this.send({ items: args, channel: logChannel || defaultChannel })),
        warn: (...args) =>
          q.push(() => this.send({ items: args, channel: warnChannel || defaultChannel })),
        error: (...args) =>
          q.push(() => this.send({ items: args, channel: errorChannel || defaultChannel })),
      },
      minLogLevel,
    )
  }
}

export function slackDefaultMessagePrefixHook(msg: SlackMessage): string[] {
  const tokens = [dayjs().toPretty()]
  const { ctx } = msg

  // AppEngine-specific decoration
  if (GAE && ctx && typeof ctx === 'object' && typeof ctx.header === 'function') {
    tokens.push(ctx.header('x-appengine-country')!, ctx.header('x-appengine-city')!)
  }

  return tokens.filter(Boolean)
}
