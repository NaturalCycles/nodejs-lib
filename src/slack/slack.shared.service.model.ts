import { StringMap } from '@naturalcycles/js-lib'
import { DebugLogLevel } from '..'

export interface SlackMessage {
  /**
   * @default bot
   */
  username?: string

  channel?: string
  icon_url?: string

  /**
   * @default :spider_web:
   */
  icon_emoji?: string

  /**
   * You can throw anything at it, it'll handle it appropriately:
   * String - as is
   * Object - pass via util.inspect()
   * Error - print the stack nicely
   */
  text: any

  level?: DebugLogLevel
  attachments?: SlackMessageAttachment[]

  /**
   * Keys-values will be rendered as MessageAttachment with Fields
   */
  kv?: StringMap<any>

  /**
   * If specified - adds @name1, @name2 in the end of the message
   */
  mentions?: string[]

  /**
   * @default false
   * By default it ignores possible errors from slack
   */
  throwOnError?: boolean

  /**
   * @default false
   * Skips logging message
   */
  noLog?: boolean
}

export interface SlackAttachmentField {
  title: string
  value: string
  short?: boolean
}

// Taken from here: https://github.com/slackapi/node-slack-sdk/blob/master/packages/types/src/index.ts
export interface SlackMessageAttachment {
  // blocks?: (KnownBlock | Block)[];
  fallback?: string // either this or text must be defined
  color?: 'good' | 'warning' | 'danger' | string
  pretext?: string
  author_name?: string
  author_link?: string // author_name must be present
  author_icon?: string // author_name must be present
  title?: string
  title_link?: string // title must be present
  text?: string // either this or fallback must be defined
  fields?: SlackAttachmentField[]
  image_url?: string
  thumb_url?: string
  footer?: string
  footer_icon?: string // footer must be present
  ts?: number
  // actions?: AttachmentAction[];
  callback_id?: string
  mrkdwn_in?: ('pretext' | 'text' | 'fields')[]
}

export interface SlackSharedServiceCfg {
  /**
   * Undefined means slack is disabled.
   */
  webhookUrl?: string

  defaults?: Partial<SlackMessage>

  /**
   * Override channel when msg.level is set.
   * key: DebugLogLevel
   * value: channel name to send message to
   */
  channelByLevel?: StringMap
}
