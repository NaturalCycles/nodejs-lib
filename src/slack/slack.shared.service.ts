import { anyToErrorObject } from '@naturalcycles/js-lib'
import { dayjs } from '@naturalcycles/time-lib'
import * as got from 'got'
import { Debug, DebugLogLevel } from '..'
import {
  SlackAttachmentField,
  SlackMessage,
  SlackSharedServiceCfg,
} from './slack.shared.service.model'

const GAE = !!process.env.GAE_INSTANCE

const DEFAULTS = (): SlackMessage => ({
  username: 'bot',
  channel: '#log',
  icon_emoji: ':spider_web:',
  text: 'no text',
})

const log = Debug('nc:nodejs-lib:slack')

export class SlackSharedService<CTX = any> {
  constructor(private slackServiceCfg: SlackSharedServiceCfg) {}

  // Convenience method
  async send(text: string, ctx?: CTX): Promise<void> {
    await this.sendMsg(
      {
        text,
      },
      ctx,
    )
  }

  /**
   * Send error.
   */
  async error(_err: any, opts: Partial<SlackMessage> = {}, ctx?: CTX): Promise<void> {
    const err = anyToErrorObject(_err)
    const text = err.stack || err.message
    await this.sendMsg(
      {
        level: DebugLogLevel.error,
        ...opts,
        text,
      },
      ctx,
    )
  }

  async sendMsg(_msg: SlackMessage, ctx?: CTX): Promise<void> {
    const { webhookUrl } = this.slackServiceCfg

    log[_msg.level || DebugLogLevel.info](...[_msg.text, _msg.kv, _msg.attachments].filter(Boolean))

    if (!webhookUrl) return

    this.processKV(_msg)

    const body: SlackMessage = {
      ...DEFAULTS(),
      ...this.slackServiceCfg.defaults,
      ..._msg,
    }

    body.channel = (this.slackServiceCfg.channelByLevel || {})[_msg.level!] || body.channel

    await this.decorateMsg(body, ctx)

    await got
      .post(webhookUrl, {
        json: true,
        body,
      })
      .catch(ignored => {}) // ignore, cause slack is weirdly returning non-json text "ok" response
  }

  /**
   * Mutates msg.
   * To be overridden.
   */
  protected async decorateMsg(msg: SlackMessage, ctx?: CTX): Promise<void> {
    const tokens = [dayjs().toPretty()]

    // AppEngine-specific decoration
    if (GAE && ctx && typeof ctx === 'object' && typeof (ctx as any).header === 'function') {
      tokens.push(
        (ctx as any).header('x-appengine-country')!,
        (ctx as any).header('x-appengine-city')!,
        // ctx.header('x-appengine-user-ip')!,
      )
    }

    msg.text = [tokens.filter(Boolean).join(': '), msg.text].join('\n')
  }

  kvToFields(kv: Record<string, any>): SlackAttachmentField[] {
    return Object.entries(kv).map(([k, v]) => ({
      title: k,
      value: String(v),
      short: String(v).length < 80,
    }))
  }

  /**
   * mutates
   */
  private processKV(msg: SlackMessage): void {
    if (!msg.kv) return

    msg.attachments = (msg.attachments || []).concat({
      fields: this.kvToFields(msg.kv),
    })

    delete msg.kv
  }
}
