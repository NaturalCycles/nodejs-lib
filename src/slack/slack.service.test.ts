import { mockTime } from '@naturalcycles/dev-lib/dist/testing'
import { noopLogger } from '@naturalcycles/js-lib'
import nock = require('nock')
import { slackDefaultMessagePrefixHook, SlackService } from './slack.service'

let lastBody: any

beforeEach(() => {
  lastBody = null
})

beforeAll(() => {
  mockTime()
})

afterAll(() => {
  nock.cleanAll()
  nock.restore() // prevents jest memory leak
})

nock(/.*/)
  .persist()
  .post(/.*/)
  .reply((uri, body: any) => {
    lastBody = body
    return uri.includes('error') ? [500, 'some error!'] : [200, 'ok']
  })

const slackService = new SlackService({
  webhookUrl: 'https://dummyhook.com',
  logger: noopLogger,
  // defaults: {
  //   channel: 'test',
  // },
})

test('basic test', async () => {
  await slackService.log('hey')
  expect(lastBody).toMatchSnapshot()

  await slackService.send('hey')
  expect(lastBody).toMatchSnapshot()

  // Array of 1 item
  await slackService.send({ items: ['yo'] })
  expect(lastBody).toMatchSnapshot()

  // Just adding "everything" for coverage/snapshot
  const err = new Error('c error!')
  delete err.stack // to not include stack in the tested snapshot (flaky)

  await slackService.send(
    {
      items: ['a', { b: 'b' }, err],
      kv: { k1: 'v1' },
      mentions: ['kirill'],
    },
    { ctxKey: 'ctxValue' },
  )
  expect(lastBody).toMatchSnapshot()

  // await expect(slackService.send({items: 'hey', throwOnError: true})).rejects.toThrow('Network request forbidden')
})

test('no webhookUrl', async () => {
  const s = new SlackService({
    logger: noopLogger,
  })
  await s.log('anything')
  expect(lastBody).toBeNull() // should be unchanged
})

test('error', async () => {
  const s = new SlackService({
    webhookUrl: 'wrongUrl',
    logger: noopLogger,
  })

  // This should NOT throw, because errors are suppressed
  await s.log('yo')

  await expect(s.send({ items: 'yo', throwOnError: true })).rejects.toThrow('Invalid URL')
})

test('messagePrefixHook returning null should NOT be sent', async () => {
  const s = new SlackService({
    webhookUrl: 'https://valid.com',
    messagePrefixHook: () => null,
    logger: noopLogger,
  })
  await s.log('yo')
  expect(lastBody).toBeNull()

  // revert to default hook
  s.cfg.messagePrefixHook = slackDefaultMessagePrefixHook
  await s.log('yo')
  expect(lastBody).not.toBeNull()
})
