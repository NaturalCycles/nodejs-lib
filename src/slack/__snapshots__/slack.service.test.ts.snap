// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic test 1`] = `
{
  "channel": "#log",
  "icon_emoji": ":spider_web:",
  "text": "2018-06-21 00:00:00
hey",
  "username": "bot",
}
`;

exports[`basic test 2`] = `
{
  "channel": "#log",
  "icon_emoji": ":spider_web:",
  "text": "2018-06-21 00:00:00
hey",
  "username": "bot",
}
`;

exports[`basic test 3`] = `
{
  "channel": "#log",
  "icon_emoji": ":spider_web:",
  "text": "2018-06-21 00:00:00
\`\`\`yo\`\`\`",
  "username": "bot",
}
`;

exports[`basic test 4`] = `
{
  "attachments": [
    {
      "fields": [
        {
          "short": true,
          "title": "k1",
          "value": "v1",
        },
      ],
    },
  ],
  "channel": "#log",
  "icon_emoji": ":spider_web:",
  "mentions": [
    "kirill",
  ],
  "text": "2018-06-21 00:00:00
\`\`\`a
{ b: 'b' }
Error: c error!\`\`\`
<@kirill>",
  "username": "bot",
}
`;
